File: sstvae/checkpoint.py
"""Locating the model checkpoint the CLI tools decode with.

The encoder/decoder network *is* the codec, so every tool needs a
checkpoint. Rather than make each user find one, the tools default to a
published checkpoint on the Hub and download it once (huggingface_hub
caches it under ~/.cache/huggingface, so it's a one-time cost).

`DEFAULT_FILE` is deliberately a fixed filename rather than a moving
"latest": the on-air format is not frozen, and a checkpoint that silently
changed under a user would break interoperability with every station
still running the old one. When a new public checkpoint is published,
bump `DEFAULT_FILE` here in the same change as the code that needs it.

Published checkpoints are therefore **immutable**, which is what lets a
cache hit be trusted outright rather than revalidated — see
`default_checkpoint`.
"""

from pathlib import Path

DEFAULT_REPO = "arodland/sstvae"
DEFAULT_FILE = "v4.pt"

# The ONNX artifacts exported from DEFAULT_FILE by scripts/export_onnx.py.
# `DEFAULT_REVISION` is the stem of the checkpoint they came from, so the
# two can never be bumped independently: change DEFAULT_FILE and the
# ONNX names follow automatically. Each artifact also carries its source
# checkpoint's sha256 in its ONNX metadata, so a mismatch is detectable
# rather than merely unlikely.
DEFAULT_REVISION = DEFAULT_FILE.rsplit(".", 1)[0]
PRECISIONS = ("fp32", "fp16", "int8")

# The decoder's gradient graph, for transmit-time latent optimization
# (`docs/latent-optimization.md`). Not a codec part: no receiver ever
# loads it, and a station that does not use the optimizer never fetches
# it.
#
# **It is fp32 whatever the codec's precision is**, because fp32 is the
# only version published -- the fp16 converter emits a graph
# onnxruntime will not load, and int8 is excluded on principle since
# differentiating `ConvInteger` is not well defined. The override is
# silent rather than an error because `--precision` is a statement about
# the *codec*, and refusing to optimize because someone picked int8 for
# their decoder would be answering a question they did not ask.
GRAD_PART = "decoder-grad"
GRAD_PRECISION = "fp32"
PARTS = ("encoder", "decoder", GRAD_PART)
# Revisions that actually have one. v1 and v2 predate the feature, so
# this is not hypothetical bookkeeping -- it is the difference between a
# clear message and a 404 for anyone pinning an older revision. Add to
# it when a revision ships the artifact.
GRAD_REVISIONS = frozenset({"v3", "v4"})

# fp16 is the shipped default: measured identical to fp32 end to end
# (docs/onnx.md) at half the size. int8 is available but costs ~1 dB of
# effective SNR on the *encoder*, whose error every receiver pays for.
DEFAULT_PRECISION = "fp16"


def onnx_filename(part: str, precision: str = DEFAULT_PRECISION) -> str:
    """e.g. ("encoder", "fp16") -> "v1-encoder-fp16.onnx".

    `decoder-grad` ignores `precision` and is always fp32 -- see
    `GRAD_PRECISION`.
    """
    if part not in PARTS:
        raise ValueError(f"part must be one of {', '.join(PARTS)}, not {part!r}")
    if precision not in PRECISIONS:
        raise ValueError(
            f"precision must be one of {', '.join(PRECISIONS)}, not {precision!r}"
        )
    if part == GRAD_PART:
        precision = GRAD_PRECISION
    return f"{DEFAULT_REVISION}-{part}-{precision}.onnx"


def default_onnx(part: str, precision: str = DEFAULT_PRECISION) -> str:
    """Path to a published ONNX artifact, downloading it once if needed.

    Same immutability argument as `default_checkpoint`: the filename
    names a specific artifact, so a cache hit needs no revalidation.

    **One part at a time, on purpose.** No CLI needs both: encoding uses
    only the encoder, decoding and listening only the decoder. Fetching
    per part means a receive-only station downloads 9 MB rather than the
    21 MB pair.
    """
    return _fetch(onnx_filename(part, precision), "onnxruntime")


def resolve_onnx(part: str, path: str | None = None,
                 precision: str = DEFAULT_PRECISION) -> str:
    """Locate one ONNX part, given whatever the user passed to --model.

    The `.pt` checkpoint is one file and the ONNX codec is two, so
    `--model` cannot simply be "the model file" any more. It accepts:

      omitted            the published artifacts (fetched, cached)
      a directory        `*-{part}-{precision}.onnx` inside it
      a `.onnx` file     that part; the sibling part is derived from the
                         name by substitution, and only when it is
                         actually needed

    A `.pt` path never reaches here -- `codec.load_codec` sends those to
    the torch backend instead.

    `decoder-grad` is forced to fp32 (see `GRAD_PRECISION`), including
    when resolving out of a directory or off a sibling.
    """
    if part == GRAD_PART:
        precision = GRAD_PRECISION
    if path is None:
        if part == GRAD_PART and DEFAULT_REVISION not in GRAD_REVISIONS:
            # The gradient graph arrived with v3; earlier revisions were
            # published without one. Say so, rather than letting the
            # fetch fail with a 404 on a filename nobody recognises --
            # the operator has done nothing wrong and the fix is a flag.
            raise SystemExit(
                f"latent optimization needs the {GRAD_PART} artifact, which "
                f"was not published for {DEFAULT_REVISION} (it exists from "
                f"{sorted(GRAD_REVISIONS)[0]} onward).\n"
                f"Pass --model with a revision that has it, e.g.\n"
                f"  --model {sorted(GRAD_REVISIONS)[0]}-encoder-"
                f"{DEFAULT_PRECISION}.onnx\n"
                "or a directory of exported .onnx files."
            )
        return default_onnx(part, precision)

    p = Path(path)
    if p.is_dir():
        matches = sorted(p.glob(f"*-{part}-{precision}.onnx"))
        if not matches:
            available = sorted(q.name for q in p.glob("*.onnx"))
            raise SystemExit(
                f"no *-{part}-{precision}.onnx in {p}\n"
                + (f"found: {', '.join(available)}" if available
                   else "that directory holds no .onnx files at all")
            )
        if len(matches) > 1:
            raise SystemExit(
                f"ambiguous: {len(matches)} candidates for {part}/{precision} "
                f"in {p} ({', '.join(q.name for q in matches)})"
            )
        return str(matches[0])

    if p.suffix == ".onnx":
        # `-decoder-` is a prefix of `-decoder-grad-`, so a plain
        # substring test would hand back the gradient graph when the
        # decoder was asked for -- and it would load, and produce
        # nonsense, because its first output is the reconstruction.
        is_grad = f"-{GRAD_PART}-" in p.name
        if (f"-{part}-" in p.name) and (is_grad == (part == GRAD_PART)):
            return str(p)
        # Named a different part: derive this one from it, so `--model
        # v1-encoder-fp16.onnx` still works for an operation that turns
        # out to need the decoder too.
        other = GRAD_PART if is_grad else (
            "decoder" if part == "encoder" else "encoder")
        if f"-{other}-" not in p.name:
            raise SystemExit(
                f"cannot tell which part {p.name} is -- expected a name like "
                f"v1-{part}-{precision}.onnx, or pass the directory instead"
            )
        # Rebuild from the stem rather than substituting the part into
        # the given name: the sibling's *precision* need not match the
        # one we were handed. `decoder-grad` is only published at fp32,
        # so `--model v3-encoder-fp16.onnx` has to resolve to
        # `v3-decoder-grad-fp32.onnx`, not to an fp16 name that was
        # never exported.
        stem = p.name.split(f"-{other}-")[0]
        sibling = p.with_name(f"{stem}-{part}-{precision}.onnx")
        if not sibling.exists():
            raise SystemExit(
                f"need the {part} as well as the {other}, but {sibling.name} "
                f"is not next to {p.name}"
            )
        return str(sibling)

    raise SystemExit(
        f"--model {path}: expected a .pt checkpoint, a .onnx artifact, or a "
        "directory containing the exported .onnx files"
    )


def default_checkpoint() -> str:
    """Path to the published checkpoint, downloading it once if needed.

    The cache is consulted **without touching the network** first, and a
    hit is returned as-is. `DEFAULT_FILE` names a specific immutable
    checkpoint (see above), so once it is cached there is nothing to
    revalidate — whereas plain `hf_hub_download` issues a HEAD request on
    every call to check for a newer version. That costs a round trip on
    each run, fails needlessly when offline, and prints a "you are
    sending unauthenticated requests to the HF Hub" warning on any
    machine without Hub credentials. For a long-running GUI that resolves
    the checkpoint at startup, none of that buys anything.
    """
    return _fetch(DEFAULT_FILE, "torch")


def _any_cached_revision(filename: str) -> str | None:
    """The file under *any* cached revision of the repo, or None.

    `hf_hub_download(local_files_only=True)` resolves `refs/main` to one
    commit and looks only inside that snapshot, so a file cached before
    an unrelated commit stops resolving offline the moment anything else
    is downloaded. That is a real failure and not a hypothetical one:
    the codec fetches the encoder and decoder **separately, on demand**,
    so downloading one can strand the other on a machine that then goes
    offline -- a field laptop being exactly the case that matters.

    Falling back like this is safe *because* published filenames are
    immutable (see the module docstring). Any revision's `v1.pt` is byte
    for byte the same `v1.pt`; the commit it happens to be filed under
    carries no information.
    """
    try:
        import huggingface_hub

        cache = Path(huggingface_hub.constants.HF_HUB_CACHE)
    except Exception:
        return None  # no real hub installed (or a stand-in, as in tests)

    snapshots = cache / f"models--{DEFAULT_REPO.replace('/', '--')}" / "snapshots"
    if not snapshots.is_dir():
        return None
    for snapshot in sorted(snapshots.iterdir(), reverse=True):
        candidate = snapshot / filename
        if candidate.exists():
            return str(candidate)
    return None


def _fetch(filename: str, needed_for: str) -> str:
    """Cache-first fetch of one immutable published file.

    Shared by the `.pt` and `.onnx` paths because the argument for
    trusting a cache hit outright is the same for both: the filename
    names a specific artifact that will never change under us.
    """
    try:
        from huggingface_hub import hf_hub_download
    except ImportError as e:  # pragma: no cover - depends on install extras
        raise SystemExit(
            f"huggingface_hub is needed to fetch {filename}.\n"
            "Install it (pip install huggingface_hub), or pass --model with a\n"
            "local .pt, .onnx, or directory of exported .onnx files."
        ) from e

    try:
        return hf_hub_download(DEFAULT_REPO, filename, local_files_only=True)
    except Exception:
        pass  # not under the current revision -- try the others, then fetch

    found = _any_cached_revision(filename)
    if found is not None:
        return found

    try:
        return hf_hub_download(DEFAULT_REPO, filename)
    except Exception as e:  # network down, Hub outage, offline machine...
        raise SystemExit(
            f"could not fetch {DEFAULT_REPO}/{filename} (needed by "
            f"{needed_for}): {e}\n"
            "If you're offline, pass --model with a local .pt, .onnx, or\n"
            "directory of exported .onnx files."
        ) from e


def resolve(path: str | None) -> str:
    """An explicit --model wins; otherwise fall back to the published one."""
    return path if path else default_checkpoint()

File: native/core/checkpoint/checkpoint.cpp
#include "checkpoint/checkpoint.hpp"

#include <algorithm>
#include <cstddef>
#include <cstdlib>
#include <filesystem>
#include <mutex>
#include <string>
#include <utility>
#include <vector>

namespace sstvae::checkpoint {

namespace fs = std::filesystem;

namespace {

std::mutex g_fetcher_mutex;
Fetcher g_fetcher;  // NOLINT(cert-err58-cpp) - plain default construction

std::optional<std::string> env(const char* name) {
    const char* v = std::getenv(name);
    if (v == nullptr || *v == '\0') return std::nullopt;
    return std::string(v);
}

fs::path home() {
#if defined(_WIN32)
    if (const auto p = env("USERPROFILE")) return fs::path(*p);
    return fs::path(".");
#else
    if (const auto p = env("HOME")) return fs::path(*p);
    return fs::path(".");
#endif
}

bool is_valid_part(std::string_view part) {
    return part == "encoder" || part == "decoder" || part == GRAD_PART;
}

std::string parts_text() { return "'encoder', 'decoder' or 'decoder-grad'"; }

// The gradient graph exists at one precision only; see GRAD_PRECISION.
std::string_view effective_precision(std::string_view part,
                                     std::string_view precision) {
    return part == GRAD_PART ? GRAD_PRECISION : precision;
}

bool revision_has_grad() {
    return std::find(GRAD_REVISIONS.begin(), GRAD_REVISIONS.end(),
                     DEFAULT_REVISION) != GRAD_REVISIONS.end();
}

bool is_valid_precision(std::string_view precision) {
    return std::find(PRECISIONS.begin(), PRECISIONS.end(), precision) !=
           PRECISIONS.end();
}

std::string precisions_text() {
    std::string out;
    for (std::size_t i = 0; i < PRECISIONS.size(); ++i) {
        if (i != 0) out += ", ";
        out += std::string(PRECISIONS[i]);
    }
    return out;
}

std::string join(const std::vector<std::string>& items, const char* sep) {
    std::string out;
    for (std::size_t i = 0; i < items.size(); ++i) {
        if (i != 0) out += sep;
        out += items[i];
    }
    return out;
}

// What to tell someone whose machine cannot reach the Hub. Named
// because this text is a deliverable rather than an afterthought: on a
// field laptop it is the whole of the recovery path.
std::string offline_advice(std::string_view filename) {
    return "\nIf you are offline, fetch it on another machine from\n  " +
           artifact_url(filename) + "\nthen either pass --model with that file, or drop it in\n  " +
           cache_dir() + "\nwhere it will be found on the next run.";
}

}  // namespace

std::string onnx_filename(std::string_view part, std::string_view precision) {
    if (!is_valid_part(part)) {
        throw CheckpointError("part must be " + parts_text() + ", not '" +
                              std::string(part) + "'");
    }
    if (!is_valid_precision(precision)) {
        throw CheckpointError("precision must be one of " + precisions_text() +
                              ", not '" + std::string(precision) + "'");
    }
    return std::string(DEFAULT_REVISION) + "-" + std::string(part) + "-" +
           std::string(effective_precision(part, precision)) + ".onnx";
}

std::string cache_dir() {
    if (const auto override_dir = env("SSTVAE_MODEL_CACHE")) return *override_dir;
#if defined(_WIN32)
    if (const auto local = env("LOCALAPPDATA")) {
        return (fs::path(*local) / "sstvae" / "cache" / "models").string();
    }
    return (home() / "AppData" / "Local" / "sstvae" / "cache" / "models").string();
#elif defined(__APPLE__)
    return (home() / "Library" / "Caches" / "sstvae" / "models").string();
#else
    if (const auto xdg = env("XDG_CACHE_HOME")) {
        return (fs::path(*xdg) / "sstvae" / "models").string();
    }
    return (home() / ".cache" / "sstvae" / "models").string();
#endif
}

std::optional<std::string> find_cached(std::string_view filename) {
    std::error_code ec;
    const fs::path p = fs::path(cache_dir()) / std::string(filename);
    if (fs::is_regular_file(p, ec)) return p.string();
    return std::nullopt;
}

std::string artifact_url(std::string_view filename) {
    return "https://huggingface.co/" + std::string(DEFAULT_REPO) + "/resolve/main/" +
           std::string(filename);
}

Fetcher default_fetcher() {
    std::lock_guard<std::mutex> lock(g_fetcher_mutex);
    return g_fetcher;
}

void set_default_fetcher(Fetcher fetcher) {
    std::lock_guard<std::mutex> lock(g_fetcher_mutex);
    g_fetcher = std::move(fetcher);
}

namespace {

// The published artifact: cache first, network only if it is missing.
std::string fetch_published(std::string_view part, std::string_view precision,
                            const Fetcher& fetcher) {
    const std::string filename = onnx_filename(part, precision);
    if (const auto cached = find_cached(filename)) return *cached;

    const Fetcher& fn = fetcher ? fetcher : default_fetcher();
    if (!fn) {
        throw CheckpointError("no " + filename + " in " + cache_dir() +
                              ", and this build cannot download it." +
                              offline_advice(filename));
    }
    try {
        return fn(filename);
    } catch (const std::exception& e) {
        // Every failure gets the advice appended, including a
        // CheckpointError from the fetcher itself. Rethrowing those
        // unchanged -- to avoid double-wrapping -- silently dropped the
        // one piece of text that tells an offline operator what to do,
        // which is this phase's actual deliverable.
        throw CheckpointError(std::string(e.what()) + offline_advice(filename));
    }
}

}  // namespace

std::string resolve_onnx(std::string_view part, std::string_view path,
                         std::string_view precision, const Fetcher& fetcher) {
    if (!is_valid_part(part)) {
        throw CheckpointError("part must be " + parts_text() + ", not '" +
                              std::string(part) + "'");
    }
    precision = effective_precision(part, precision);
    if (path.empty()) {
        if (part == GRAD_PART && !revision_has_grad()) {
            throw CheckpointError(
                "latent optimization needs the " + std::string(GRAD_PART) +
                " artifact, which was not published for " +
                std::string(DEFAULT_REVISION) + " (it exists from " +
                std::string(GRAD_REVISIONS.front()) +
                " onward).\nPass --model with a revision that has it, or a "
                "directory of exported .onnx files.");
        }
        return fetch_published(part, precision, fetcher);
    }

    std::error_code ec;
    const fs::path p{std::string(path)};

    if (fs::is_directory(p, ec)) {
        const std::string suffix =
            "-" + std::string(part) + "-" + std::string(precision) + ".onnx";
        std::vector<std::string> matches;
        std::vector<std::string> any_onnx;
        for (const fs::directory_entry& entry : fs::directory_iterator(p, ec)) {
            const std::string name = entry.path().filename().string();
            if (name.size() < 5 || name.compare(name.size() - 5, 5, ".onnx") != 0) {
                continue;
            }
            any_onnx.push_back(name);
            if (name.size() > suffix.size() &&
                name.compare(name.size() - suffix.size(), suffix.size(), suffix) == 0) {
                matches.push_back(entry.path().string());
            }
        }
        std::sort(matches.begin(), matches.end());
        std::sort(any_onnx.begin(), any_onnx.end());

        if (matches.empty()) {
            throw CheckpointError(
                "no *-" + std::string(part) + "-" + std::string(precision) +
                ".onnx in " + p.string() + "\n" +
                (any_onnx.empty() ? "that directory holds no .onnx files at all"
                                  : "found: " + join(any_onnx, ", ")));
        }
        if (matches.size() > 1) {
            std::vector<std::string> names;
            names.reserve(matches.size());
            for (const std::string& m : matches) {
                names.push_back(fs::path(m).filename().string());
            }
            throw CheckpointError("ambiguous: " + std::to_string(matches.size()) +
                                  " candidates for " + std::string(part) + "/" +
                                  std::string(precision) + " in " + p.string() + " (" +
                                  join(names, ", ") + ")");
        }
        return matches.front();
    }

    if (p.extension() == ".onnx") {
        const std::string name = p.filename().string();
        const std::string mine = "-" + std::string(part) + "-";
        const std::string grad_tag = "-" + std::string(GRAD_PART) + "-";
        // `-decoder-` is a prefix of `-decoder-grad-`, so a bare
        // substring test hands back the gradient graph when the decoder
        // was asked for -- and it loads, and its first output *is* a
        // reconstruction, so the mistake survives to a wrong picture
        // rather than an error.
        const bool is_grad = name.find(grad_tag) != std::string::npos;
        if (name.find(mine) != std::string::npos && is_grad == (part == GRAD_PART)) {
            return p.string();
        }

        // Named a different part: derive this one, so `--model
        // v1-encoder-fp16.onnx` still works for an operation that turns
        // out to need the decoder too.
        const std::string other_part =
            is_grad ? std::string(GRAD_PART)
                    : (part == "encoder" ? "decoder" : "encoder");
        const std::string theirs = "-" + other_part + "-";
        const std::size_t at = name.find(theirs);
        if (at == std::string::npos) {
            throw CheckpointError("cannot tell which part " + name +
                                  " is -- expected a name like " +
                                  std::string(DEFAULT_REVISION) + "-" +
                                  std::string(part) + "-" + std::string(precision) +
                                  ".onnx, or pass the directory instead");
        }
        // Rebuild from the stem rather than substituting into the given
        // name: the sibling's *precision* need not match the one we
        // were handed. `decoder-grad` is published at fp32 only, so the
        // gradient sibling of an fp16 encoder is an fp32 name that a
        // substitution would never produce.
        const std::string sibling_name = name.substr(0, at) + mine +
                                         std::string(precision) + ".onnx";
        const fs::path sibling = p.parent_path() / sibling_name;
        if (!fs::is_regular_file(sibling, ec)) {
            throw CheckpointError("need the " + std::string(part) + " as well as the " +
                                  other_part + ", but " + sibling_name +
                                  " is not next to " + name);
        }
        return sibling.string();
    }

    throw CheckpointError("--model " + std::string(path) +
                          ": expected a .onnx artifact or a directory containing "
                          "the exported .onnx files");
}

}  // namespace sstvae::checkpoint
